File: macos/Sources/VoiceKeyCore/VocabularyDictionary.swift
import Foundation

/// User-editable custom vocabulary: terms that bias Whisper through an initial prompt,
/// plus spoken-to-written replacement rules applied to the finished transcript.
public struct VocabularyDictionary: Codable, Equatable {
    public var terms: [String]
    public var replacements: [String: String]

    public init(terms: [String] = [], replacements: [String: String] = [:]) {
        self.terms = terms
        self.replacements = replacements
    }

    /// Whisper initial-prompt text biasing recognition toward the terms, or nil when there are no terms.
    public var promptText: String? {
        let named = terms.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        guard !named.isEmpty else { return nil }
        return "Glossary: " + named.joined(separator: ", ") + "."
    }

    /// Applies every replacement rule in a single left-to-right pass, so a rule's output
    /// can never be rewritten by another rule. Longer keys win over overlapping shorter ones.
    public func applyingReplacements(to text: String) -> String {
        guard let matcher = Matcher(replacements: replacements) else { return text }
        return matcher.applied(to: text)
    }

    /// Starter content written on first use so the user edits a documented example.
    public static let template = VocabularyDictionary(
        terms: ["VoiceKey", "WhisperKit", "Kubernetes"],
        replacements: [
            "get hub": "GitHub",
            "jason": "JSON",
        ]
    )

    /// A transcript offered back as a vocabulary term, or nil when it is plainly
    /// a sentence rather than a name: terms bias recognition, and biasing towards
    /// a whole sentence would do more harm than good.
    public static func suggestedTerm(from transcript: String) -> String? {
        let words = transcript.split(whereSeparator: { $0.isWhitespace })
        guard (1...4).contains(words.count) else { return nil }
        let term = words.joined(separator: " ")
        guard term.count <= 40 else { return nil }
        return term
    }

    /// Reads JSON from disk; nil if the file is missing or malformed — a broken user edit
    /// must degrade to "no custom vocabulary", never crash the app.
    public static func load(from url: URL) -> VocabularyDictionary? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(VocabularyDictionary.self, from: data)
    }

    public func save(to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(self).write(to: url, options: .atomic)
    }
}

/// One combined case-insensitive alternation over every rule key. Building a single
/// expression is what guarantees the single pass; ordering the alternatives longest-first
/// is what gives longer keys precedence.
private struct Matcher {
    private let expression: Regex<AnyRegexOutput>
    private let valuesByLowercasedKey: [String: String]

    init?(replacements: [String: String]) {
        let keys = replacements.keys
            .filter { !$0.isEmpty }
            .sorted { ($0.count, $0) > ($1.count, $1) }
        guard !keys.isEmpty else { return nil }

        let pattern = keys.map(Matcher.wholeWordPattern).joined(separator: "|")
        guard let expression = try? Regex(pattern).ignoresCase() else { return nil }
        self.expression = expression
        self.valuesByLowercasedKey = replacements.reduce(into: [:]) { result, entry in
            result[entry.key.lowercased()] = entry.value
        }
    }

    func applied(to text: String) -> String {
        text.replacing(expression) { match in
            let matched = String(match.output[0].substring ?? "")
            return valuesByLowercasedKey[matched.lowercased()] ?? matched
        }
    }

    /// Escapes the key so metacharacters ("c++", "node.js") stay literal, and fences it with
    /// boundaries that adapt to the edge characters — `\b` only works next to word characters,
    /// so keys ending in punctuation get a "not followed by a word character" assertion instead.
    private static func wholeWordPattern(for key: String) -> String {
        let leading = key.first.map(isWordCharacter) == true ? #"\b"# : #"(?<!\w)"#
        let trailing = key.last.map(isWordCharacter) == true ? #"\b"# : #"(?!\w)"#
        return leading + NSRegularExpression.escapedPattern(for: key) + trailing
    }

    private static func isWordCharacter(_ character: Character) -> Bool {
        character == "_" || character.isLetter || character.isNumber
    }
}

File: windows/VoiceKey.Core/VocabularyDictionary.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace VoiceKey.Core;

/// <summary>
/// User-editable custom vocabulary: terms that bias Whisper through an initial prompt,
/// plus spoken-to-written replacement rules applied to the finished transcript.
/// </summary>
public sealed class VocabularyDictionary : IEquatable<VocabularyDictionary>
{
    public IReadOnlyList<string> Terms { get; }
    public IReadOnlyDictionary<string, string> Replacements { get; }

    public VocabularyDictionary(
        IReadOnlyList<string>? terms = null,
        IReadOnlyDictionary<string, string>? replacements = null)
    {
        Terms = terms ?? [];
        Replacements = replacements ?? new Dictionary<string, string>();
    }

    /// <summary>
    /// Whisper initial-prompt text biasing recognition toward the terms, or null when
    /// there are no terms.
    /// </summary>
    public string? PromptText
    {
        get
        {
            var named = Terms.Where(term => !string.IsNullOrWhiteSpace(term)).ToArray();
            return named.Length == 0 ? null : "Glossary: " + string.Join(", ", named) + ".";
        }
    }

    /// <summary>
    /// Applies every replacement rule in a single left-to-right pass, so a rule's output
    /// can never be rewritten by another rule. Longer keys win over overlapping shorter ones.
    /// </summary>
    public string ApplyingReplacements(string text) =>
        Matcher.For(Replacements)?.Applied(text) ?? text;

    /// <summary>Starter content written on first use so the user edits a documented example.</summary>
    public static VocabularyDictionary Template { get; } = new(
        terms: ["VoiceKey", "Whisper", "Kubernetes"],
        replacements: new Dictionary<string, string>
        {
            ["get hub"] = "GitHub",
            ["jason"] = "JSON",
        });

    /// <summary>
    /// A transcript offered back as a vocabulary term, or null when it is plainly
    /// a sentence rather than a name: terms bias recognition, and biasing towards
    /// a whole sentence would do more harm than good.
    /// </summary>
    public static string? SuggestedTerm(string transcript)
    {
        var words = transcript.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (words.Length is < 1 or > 4) return null;
        var term = string.Join(" ", words);
        return term.Length <= 40 ? term : null;
    }

    // MARK: - Storage

    /// <summary>
    /// Reads JSON from disk; null if the file is missing or malformed — a broken user edit
    /// must degrade to "no custom vocabulary", never crash the app.
    /// </summary>
    public static VocabularyDictionary? Load(string path)
    {
        try
        {
            var dto = JsonSerializer.Deserialize<Dto>(File.ReadAllText(path));
            return dto is null ? null : new VocabularyDictionary(dto.Terms, dto.Replacements);
        }
        catch (Exception exception) when (exception is IOException or JsonException
            or UnauthorizedAccessException)
        {
            return null;
        }
    }

    public void Save(string path)
    {
        // Sorted keys and indentation so a hand-edited file stays readable and
        // diffs stay stable.
        var dto = new Dto(new SortedDictionary<string, string>(
                Replacements.ToDictionary(entry => entry.Key, entry => entry.Value),
                StringComparer.Ordinal),
            Terms);
        File.WriteAllText(path, JsonSerializer.Serialize(dto, SerializerOptions));
    }

    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        WriteIndented = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    private sealed record Dto(
        [property: JsonPropertyName("replacements")] IReadOnlyDictionary<string, string> Replacements,
        [property: JsonPropertyName("terms")] IReadOnlyList<string> Terms);

    // MARK: - Equality

    public bool Equals(VocabularyDictionary? other) =>
        other is not null
        && Terms.SequenceEqual(other.Terms)
        && Replacements.Count == other.Replacements.Count
        && Replacements.All(entry => other.Replacements.TryGetValue(entry.Key, out var value)
                                     && value == entry.Value);

    public override bool Equals(object? obj) => Equals(obj as VocabularyDictionary);

    public override int GetHashCode() => HashCode.Combine(Terms.Count, Replacements.Count);
}

/// <summary>
/// One combined case-insensitive alternation over every rule key. Building a single
/// expression is what guarantees the single pass; ordering the alternatives longest-first
/// is what gives longer keys precedence.
/// </summary>
internal sealed class Matcher
{
    private readonly Regex _expression;
    private readonly IReadOnlyDictionary<string, string> _valuesByLowercasedKey;

    private Matcher(Regex expression, IReadOnlyDictionary<string, string> valuesByLowercasedKey)
    {
        _expression = expression;
        _valuesByLowercasedKey = valuesByLowercasedKey;
    }

    internal static Matcher? For(IReadOnlyDictionary<string, string> replacements)
    {
        var keys = replacements.Keys
            .Where(key => key.Length > 0)
            .OrderByDescending(key => key.Length)
            .ThenByDescending(key => key, StringComparer.Ordinal)
            .ToArray();
        if (keys.Length == 0) return null;

        var pattern = string.Join("|", keys.Select(WholeWordPattern));
        var values = new Dictionary<string, string>();
        foreach (var entry in replacements) values[entry.Key.ToLowerInvariant()] = entry.Value;

        try
        {
            return new Matcher(new Regex(pattern, RegexOptions.IgnoreCase), values);
        }
        catch (ArgumentException)
        {
            return null;
        }
    }

    internal string Applied(string text) =>
        _expression.Replace(text, match =>
            _valuesByLowercasedKey.TryGetValue(match.Value.ToLowerInvariant(), out var value)
                ? value
                : match.Value);

    /// <summary>
    /// Escapes the key so metacharacters ("c++", "node.js") stay literal, and fences it with
    /// boundaries that adapt to the edge characters — <c>\b</c> only works next to word
    /// characters, so keys ending in punctuation get a "not followed by a word character"
    /// assertion instead.
    /// </summary>
    private static string WholeWordPattern(string key)
    {
        var leading = IsWordCharacter(key[0]) ? @"\b" : @"(?<!\w)";
        var trailing = IsWordCharacter(key[^1]) ? @"\b" : @"(?!\w)";
        return leading + Regex.Escape(key) + trailing;
    }

    private static bool IsWordCharacter(char character) =>
        character == '_' || char.IsLetter(character) || char.IsNumber(character);
}
